### PR TITLE
cptbox: don't allow paths that don't exist

### PR DESCRIPTION
**Pull request by Riolku**
This is mostly for compatibility with landlock, which enforces this. Builds on top of #999.

### CODE REVIEW

File: dmoj/executors/compiled_executor.py [hunk 1]
```diff
@@ -95,6 +98,14 @@ def create_files(self, problem_id: str, source_code: bytes, *args, **kwargs) ->
         with open(self._code, 'wb') as fo:
             fo.write(utf8bytes(source_code))
 
+        for path in self.compiler_required_dirs:
+            path = os.path.expanduser(path)
+            assert path == os.path.abspath(path), 'compiler_required_dirs must contain only absolute normalized paths'
```
Comment (Xyene) [inline dmoj/executors/compiled_executor.py:103]: Wait but... they don't? `~/.cache` isn't an absolute path. I don't think it's worth having any restriction here.

**Comment by Xyene** [comment]
Are you sure Rust doesn't need a cache directory? What does it do when it doesn't have one?

File: dmoj/executors/compiled_executor.py (hunk 1 above)
Comment (Riolku) [reply dmoj/executors/compiled_executor.py:103]: What happens if the path is not absolute after `expanduser`? What do we do?

Comment (Xyene) [reply dmoj/executors/compiled_executor.py:103]: It's fine, these are not adversarial conditions. We specify the paths.

**Comment by Riolku** [comment]
Do you mean `~/.cargo`? I'm not sure. I wonder if that's created on Rust installation or something. I'll add it to the list anyway.

File: dmoj/executors/compiled_executor.py (hunk 1 above)
Comment (Riolku) [reply dmoj/executors/compiled_executor.py:103]: Should I change the error message? I'm very hesitant to allow non-absolute paths.

**Comment by Xyene** [comment]
I don't think so, but it's worth checking? I know the first compile during autoconf takes a long time, because it has to download a bunch of crates that then get cached.

File: dmoj/executors/compiled_executor.py (hunk 1 above)
Comment (Xyene) [reply dmoj/executors/compiled_executor.py:103]: Why? What are you protecting against? This code reads to me like having a check for the sake of having a check, it doesn't provide utility.

**Comment by Riolku** [comment]
RUST passes in CI under this patch. I'm not really sure what it's doing though, and I was having trouble testing this locally. I think docker is re-using my old volumes, or something. Do you have any tips for how I should go about testing this?

File: dmoj/executors/compiled_executor.py (hunk 1 above)
Comment (Riolku) [reply dmoj/executors/compiled_executor.py]: I've changed the error message.

Comment (Riolku) [reply dmoj/executors/compiled_executor.py]: We have sanity checks in our FilesystemAccessRule, right? Why not here?

Comment (Xyene) [reply dmoj/executors/compiled_executor.py]: FilesystemAccessRule can be specified by users who may not know what they're doing in `extra_fs`, these can't

Comment (Riolku) [reply dmoj/executors/compiled_executor.py]: I've removed the check, but I'm still unsure where the line is for making these checks. Don't we specify the paths for the rules too? I suppose not _always_.

**Comment by Xyene** [comment]
Try running `docker prune -a` then rerunning `make judge-tier${whatevertierrustisin}`.

**Comment by Riolku** [comment]
Rust actually has `.cargo` setup in Docker container initialization, so we're ok.

**Comment by Riolku** [comment]
Actually, if we aren't in Docker, we should create .cargo right? Is it worth adding then?

**Comment by Xyene** [comment]
I don't want to particularly support running outside of Docker, it's too much of a hassle. Users are on their own if they do so anyway.

**Comment by Xyene** [comment]
Specifically: Cargo installs to `.cargo`, so if they don't have `.cargo` they probably don't have Cargo installed (or have it at some weird path). Not our problem in either case.

**Review by Xyene — APPROVED** [review]
LGTM

**Pull request merged**